ooasp/run.py

```diff
@@ -1,314 +1,566 @@
+        """
+        self.log("\t+++++ global_lb")
+        cautious = self.get_cautious()
+        for s in cautious:
+            if s.match("global_lb", 3):
+                self.log("\t  ---> Apply ", s)
+                c1, needed, _ = s.arguments
+                for _ in range(0, needed.number):
+                    self.add_object(c1.name)
+                return True
+        return False
+
+    def association_needed(self) -> bool:
+        """
+        The appearance of predicate association_needed(ASSOC, ID1, ID2, new_object)
+        in the brave consequences indicates the need to add an association between two objects.
+
+        We know that ID1 needs at least one object of class C1 and ID2 needs at least one object of class C2
+        We know that the objects ID1 and ID2 can potentially be associated by ASSOC
+        We also know that the classes of these objects was set byt the user or the smart association.
+        This makes sure that the association added does not determine the classes of the objects if not previously given
+
+        Returns:
+            bool: True if associations were added, False otherwise
+        """
+        self.log("\t+++++ association_needed")
+        brave = self.get_brave()
+        for s in brave:
+            if s.match("association_needed", 4):
+                self.log("\t  ---> Apply ", s)
+                assoc, id1, id2, _ = s.arguments
+                a = (str(assoc), id1, id2)
+                self.add_association(a)
+                return True
+        return False
+
+    def run(self):
+        """
+        Runs the solver.
+        It starts by creating the initial objects and then iterates over the smart generation and solving steps.
+        It stops when a solution is found or the timeout is reached.
+        If the view is enabled, it saves the solution as a PNG.
+        """
+        run_start = time.time()
+        self.create_initial_objects()
+        self.times["initialization"] = time.time() - run_start
+        done = False
+
+        def on_model(m):
+            self.model = [str(s) + "." for s in m.symbols(atoms=True)]
+
+        round = 0
+        while not done and not self.timeout_reached:
+            round += 1
+            self.log("\n" + title(f"Next round: {self.next_id - 1} objects"))
+            if self.view:
+                self.save_png("out/solve", f"-O{self.next_id-1}-round{round}")
+            start = time.time()
+            things_done = self.smart_generation()
+            self.times["smart_generation"]["time"] += time.time() - start
+            if things_done:
+                continue
+            self.log(subtitle(f"Solving for size {self.next_id - 1}...", "RED"))
+            self.ctl.configuration.solve.models = "1"
+            start = time.time()
+            with self.ctl.solve(
+                assumptions=self.assumptions, async_=True, on_model=on_model
+            ) as hdl:
```

Comment:
```suggestion
            ) as hdl:
            self.timeout = float('inf') if self.timeout is None else self.timeout
```

Reply:
This change should avoid cases, where `--timeout` option not being specified causes a crash, but this change could also be made to the default `OOASPRacksSolver.timeout` value.

Reply:
I fixed it by adding a default value to `OOASPRacksSolver.timeout`